---
title: 'Population structure analysis with ACP'
date: "`r Sys.Date()`"
---

```{r}
#| label: setup
#| include: false
knitr::opts_chunk$set(
  eval = TRUE,
  collapse = TRUE,
  include = TRUE
)
```

# Refence material and dedicated packages

The analysis workflow on this page is in large part inspired from the material provided by the Niklaus J. GrÃ¼nwald lab in [Population genetics and genomics in R](https://grunwaldlab.github.io/Population_Genetics_in_R/index.html)

Just like with any topic in R and other programming languages there are tons of over options to conduct this type of analysis. For example [`pegas`](https://emmanuelparadis.github.io/pegas.html) is another R package dedicated to population genetic analysis.

# Attach necessary packages

```{r}
#| label: attach packages

library("adegenet")
library("vcfR")
```

If they are not already available on your system. Can you take care of the installation?

```{r}
#| label: install packages
#| eval: false
#| code-fold: true
#| code-summary: "ðŸ’­ !! Check the answer !! ðŸª‚"
#| output: false


install.packages(c("adegenet", "vcfR"))
# A more 
pak::pkg_install(c("adegenet", "vcfR"))
```

# Loading the vcf file data in R and extract a snp matrix

```{r}
#| label: vcf file path
vcfPath <- here::here(".data", "all_samples.vcf.gz")
```

```{r}
#| label: processing vfc to adegenet genlight object
#| results: "markup"
#| eval: false
#| output: false
vcf <- read.vcfR(vcfPath, verbose = FALSE)
glight <- vcfR2genlight(vcf)
saveRDS(glight, file = here::here(".local", "genotypes_genlight.rds"))
```

# Run the PCA and plot

```{r}
#| label: PCA and plotting
#| fig-width: 4
#| fig-height: 5
#| fig-align: "left"
#| fig.cap: "Genotypes in the PCA subspace"

if (!exists("glight")) glight <- readRDS(here::here(".data", "genotypes_genlight.rds"))
nInd(glight)
nLoc(glight)
pca1 <- glPca(glight, nf = 20, parallel = TRUE, n.cores = 6)
print(pca1)

scatter(pca1, posi="topright")
```

The reason we picked `glPca()` is because it is specialized for genotyping data. BUT **keep in mind** that the PCA analysis could have been done a number of ways, notably using base R `stats::princomp()` or the `FactoMineR` package (excellent [MOOC](http://factominer.free.fr/course/MOOC.html) on multivariate analysis). If you want to learn more about PCA in R, we advise you to examine specialized resources describing how to do statistical analysis with R using best practices. Here are a few resources as a starter:

-   [The Handbook of Biological Statistics](https://rcompanion.org/rcompanion)
-   [datacamp](https://www.datacamp.com/doc/r/anova)
